drivers/fs.go

```diff
@@ -0,0 +1,228 @@
+	ended  bool
+	dCache map[string]*dataCache
+	dLock  sync.RWMutex
+}
+
+func NewFSDriver(baseURI *url.URL) *FSOS {
+	return &FSOS{
+		baseURI:  baseURI,
+		sessions: make(map[string]*FSSession),
+		lock:     sync.RWMutex{},
+	}
+}
+
+func (ostore *FSOS) NewSession(path string) OSSession {
+	ostore.lock.Lock()
+	defer ostore.lock.Unlock()
+	if session, ok := ostore.sessions[path]; ok {
+		return session
+	}
+	session := &FSSession{
+		os:     ostore,
+		path:   path,
+		dCache: make(map[string]*dataCache),
+		dLock:  sync.RWMutex{},
+	}
+	ostore.sessions[path] = session
+	return session
+}
+
+func (ostore *FSOS) GetSession(path string) *FSSession {
+	ostore.lock.Lock()
+	defer ostore.lock.Unlock()
+	if session, ok := ostore.sessions[path]; ok {
+		return session
+	}
+	return nil
+}
+
+func (ostore *FSOS) UriSchemes() []string {
+	return []string{""}
+}
+
+func (ostore *FSOS) Description() string {
+	return "File system driver."
+}
+
+func (ostore *FSSession) OS() OSDriver {
+	return ostore.os
+}
+
+// EndSession clears memory cache
+func (ostore *FSSession) EndSession() {
+	ostore.dLock.Lock()
+	ostore.ended = true
+	for k := range ostore.dCache {
+		delete(ostore.dCache, k)
+	}
+	ostore.dLock.Unlock()
+
+	ostore.os.lock.Lock()
+	delete(ostore.os.sessions, ostore.path)
+	ostore.os.lock.Unlock()
+}
+
+func (ostore *FSSession) ListFiles(ctx context.Context, dir, delim string) (PageInfo, error) {
+	pi := &singlePageInfo{
+		files:       []FileInfo{},
+		directories: []string{},
+	}
+	ostore.dLock.RLock()
+	defer ostore.dLock.RUnlock()
+
+	fullPath := ostore.getAbsoluteURI(dir)
+
+	if fullPath == "" {
+		return pi, nil
+	}
+
+	// list files in specified dir
+	files, err := ioutil.ReadDir(fullPath)
+	if err != nil {
+		return nil, err
+	}
+	// create metadata
+	for _, f := range files {
+		if f.IsDir() {
+			pi.directories = append(pi.directories, f.Name())
+		} else {
+			size := f.Size()
+			pi.files = append(pi.files, FileInfo{
+				Name:         f.Name(),
+				ETag:         "",
+				LastModified: f.ModTime(),
+				Size:         &size,
+			})
+		}
+	}
+	return pi, nil
+}
+
+func (ostore *FSSession) ReadData(ctx context.Context, name string) (*FileInfoReader, error) {
+	data := ostore.GetData(name)
+	if data == nil {
+		return nil, errors.New("Not found")
+	}
+	size := int64(len(data))
+	res := &FileInfoReader{
+		FileInfo: FileInfo{
+			Name: name,
+			Size: &size,
+		},
+		Body: ioutil.NopCloser(bytes.NewReader(data)),
+	}
+	return res, nil
+}
+
+// GetData returns the contents of the file. It reads entire file into memory.
+//
+// A name is relative to driver's baseUri.
+func (ostore *FSSession) GetData(name string) []byte {
+	prefix := ""
+	if ostore.os.baseURI != nil {
+		prefix += ostore.os.baseURI.String()
+	}
+	fullPath := path.Join(prefix, name)
+	file, err := os.Open(fullPath)
+	if err != nil {
+		return nil
+	}
+	defer file.Close()
+	fileData, err := ioutil.ReadAll(file)
+	if err != nil {
+		return nil
+	}
+	return fileData
```

Comment:
Would it be possible to **not** buffer the entire file in memory? Locally, it seems like we could pretty simply return an `io.ReadCloser` (`file`) from here and then just return it as the `FileInfoReader.Body`. It is the responsibility of the caller of `ReadData` to close the `Body`, and so we avoid buffering every file in memory here. Or do we need that for any reason?